errors/returning-response-body-in-middleware.md

```diff
@@ -1,86 +1,65 @@
-    const [, variantId] = cookie.split('.')
-    if (variantId !== '0') {
-      const url = req.nextUrl.clone()
-      url.pathname = url.pathname.replace('/', `/${variantId}/`)
-      // rewrites the response to display desired variant
-      res = NextResponse.rewrite(url)
-    }
+```js
+new Response('a text value')
+new Response(streamOrBuffer)
+new Response(JSON.stringify(obj), { headers: 'application/json' })
+NextResponse.json()
+```
 
-    // don't forget to set cookie if not set yet
-    if (!req.cookies.has(COOKIE_NAME)) {
-      res.cookies.set(COOKIE_NAME, cookie)
-    }
-    return res
-  }
-  ```
+### How to upgrade
 
-- authentication, where you **_redirect_** to your log-in/sign-in page any un-authenticated request
+For cases where Middleware is used to respond (such as authorization), you should migrate to use `rewrite`/`redirect` to pages that show an authorization error, login forms, or to an API Route.
 
-  ```js
-  export function middleware(req: NextRequest) {
-    const basicAuth = req.headers.get('authorization')
+#### Before
 
-    if (basicAuth) {
-      const auth = basicAuth.split(' ')[1]
-      const [user, pwd] = atob(auth).split(':')
-      if (areCredentialsValid(user, pwd)) {
-        return NextResponse.next()
-      }
-    }
+```typescript
+// pages/_middleware.ts
+import { NextResponse } from 'next/server'
+import type { NextRequest } from 'next/server'
+import { isAuthValid } from './lib/auth'
 
-    return NextResponse.redirect(
-      new URL(`/login?from=${req.nextUrl.pathname}`, req.url)
-    )
-  }
-  ```
-
-- detecting bots and **_rewrite_** response to display to some sink
-
-  ```js
-  export function middleware(req: NextRequest) {
-    if (isABotRequest(req)) {
-      // Bot detected! rewrite to the sink
-      const url = req.nextUrl.clone()
-      url.pathname = '/bot-detected'
-      return NextResponse.rewrite(url)
-    }
+export function middleware(request: NextRequest) {
+  // Example function to validate auth
+  if (isAuthValid(request)) {
     return NextResponse.next()
   }
-  ```
-
-- programmatically adding **_headers_** to the response, like cookies.
-
-  ```js
-  export function middleware(req: NextRequest) {
-    const res = NextResponse.next(null, {
-      // sets a custom response header
-      headers: { 'response-greetings': 'Hej!' },
-    })
-    // configures cookies
-    res.cookies.set('hello', 'world')
-    return res
+
+  return NextResponse.json({ message: 'Auth required' }, { status: 401 })
+}
+```
+
+#### After
+
+```typescript
+// middleware.ts
+import { NextResponse } from 'next/server'
+import type { NextRequest } from 'next/server'
+import { isAuthValid } from './lib/auth'
+
+export function middleware(request: NextRequest) {
+  // Example function to validate auth
+  if (isAuthValid(request)) {
+    return NextResponse.next()
   }
-  ```
+
+  const loginUrl = new URL('/login', request.url)
+  loginUrl.searchParams.set('from', request.nextUrl.pathname)
```

Comment:
```suggestion
  request.nextUrl.searchParams.set('from', request.nextUrl.pathname)
  request.nextUrl.pathname = '/login'
```
Creating a new URL instance breaks the URL normalizing that `nextUrl` provides so we should make sure to recommend using `nextUrl` by default.